invenio_records_permissions/generators.py

```diff
@@ -103,6 +106,71 @@ def query_filter(self, **kwargs):
         return ~dsl.Q("match_all")
 
 
+class DisableIf(Disable):
+    """Conditionally deny ALL users (including those who have superuser-access)."""
+
+    def __init__(self, condition=True):
+        """Constructor, setting the condition under which to deny access.
+
+        :param condition: The condition under which access should be denied.
+                          This can be either a constant boolean value, or a function
+                          that accepts keyword arguments and returns a boolean value.
+        """
+        self._check_condition = None
+        self._condition = condition
+
+    @property
+    def check_condition(self):
+        """Check if the set condition is true with the given kwargs."""
+        if self._check_condition is None:
+
+            def _condition_wrapper(**kwargs):
+                """Wrapper for an uncallable condition value to make it callable."""
+                return self._condition
+
+            if callable(self._condition):
+                self._check_condition = self._condition
+            else:
+                self._check_condition = _condition_wrapper
+
+        return self._check_condition
+
+    def excludes(self, **kwargs):
+        """Disallow any_user, if the condition is met."""
+        if self.check_condition(**kwargs):
+            return [any_user]
+        else:
+            return []
+
+    def query_filter(self, **kwargs):
+        """Filter everything from searches, if the condition is met."""
+        if self.check_condition(**kwargs):
+            return ~dsl.Q("match_all")
+        else:
+            return []
+
+
+class DisableIfReadOnly(DisableIf):
+    """Disable ALL users if the system is in read-only mode.
+
+    This generator uses the ``RECORDS_PERMISSIONS_READ_ONLY`` configuration item
+    to determine if the read-only mode is set.
+    """
+
+    def __init__(self):
+        """Constructor."""
+        self._check_condition = None
+
+    @property
+    def _condition(self):
+        """Check the current app's config if the read-only mode is active."""
+        # note: this isn't executed in the constructor because it can cause issues
+        # when instantiating the permission policies without app context
+        # e.g. in some cases of conftest.py
+        config_value = current_app.config.get("RECORDS_PERMISSIONS_READ_ONLY", False)
```

Comment:
question: it feels strange to do it in this function with this argument. do you remember the real issue?

Reply:
sorry, i don't think i fully understand the question but maybe some of this context helps:

the permission policies use object instances rather than classes in their properties (see e.g. [here](https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/services/permissions.py#L59)), and these could sometimes be constructed outside of an application context.
as such, we can't use `current_app` in the constructor of the generator (i forgot whether that was an issue in `invenio shell`, `invenio` CLI commands or anywhere else – according to the comment it would've at least created issues in the tests though).
instead, i went for evaluating the `current_app.config` only when the permission policy is actually used on some object and caching that config value.
the `self` argument is just there because python wants it on methods.